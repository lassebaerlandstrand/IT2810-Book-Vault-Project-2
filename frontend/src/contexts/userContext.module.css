.centeredOnPage {
  height: 100vh;
}
