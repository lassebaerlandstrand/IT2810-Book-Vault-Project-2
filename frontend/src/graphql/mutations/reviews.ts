import { gql } from '@/generated';

export const CREATE_REVIEW = gql(`
  mutation CreateReview($input: CreateReviewInput!) {
    createReview(input: $input) {
      id
      rating
      numRatings
    }
  }
`);

export const UPDATE_REVIEW = gql(`
  mutation UpdateReview($input: UpdateReviewInput!) {
    updateReview(input: $input) {
      id
      rating
      numRatings
    }
  }
`);
