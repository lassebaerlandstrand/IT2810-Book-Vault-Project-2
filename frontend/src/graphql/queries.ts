import { gql } from '@apollo/client';

export const GET_BOOKS = gql`
  query GetBooks($limit: Int, $offset: Int) {
    books(limit: $limit, offset: $offset) {
      id
      title
      authors
      publisher
      genres
      coverImg
    }
  }
`;
