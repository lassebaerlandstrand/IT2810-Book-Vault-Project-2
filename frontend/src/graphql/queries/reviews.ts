// This file defines all queries related to reviews
