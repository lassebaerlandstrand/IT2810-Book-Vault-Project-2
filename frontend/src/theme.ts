import {
  Container,
  createTheme,
  DEFAULT_THEME,
  Loader,
  MantineColorsTuple,
  rem,
} from '@mantine/core';
import { BookLoader } from './components/Loading/Loading';

const CONTAINER_SIZES: Record<string, string> = {
  xxs: '24em',
  xs: '30em',
  sm: '48em',
  md: '64em',
  lg: '75em',
  xl: '90em',
};

// https://colorkit.co/color-palette-generator/02070b-030a11-030d16-051321-06192b-092641-0c3256-0f3f6c-124b81-155896/
const darkBlue: MantineColorsTuple = [
  '#155896',
  '#124b81',
  '#0f3f6c',
  '#0c3256',
  '#092641',
  '#06192b',
  '#051321',
  '#030d16',
  '#030a11',
  '#02070b',
];

// https://colorkit.co/color-palette-generator/6a5f4a-d4bd94-dac5a1-e2d1b5-e9ddc8-eee4d3-f2eade-f6f0e7-f9f5ef-fcfaf7/
const beige: MantineColorsTuple = [
  '#fcfaf7',
  '#f9f5ef',
  '#f6f0e7',
  '#f2eade',
  '#eee4d3',
  '#e9ddc8',
  '#e2d1b5',
  '#dac5a1',
  '#d4bd94',
  '#6a5f4a',
];

export const theme = createTheme({
  fontFamily: `Poppins, ${DEFAULT_THEME.fontFamily}`,
  breakpoints: {
    xxs: '24em', // 384px
    xs: '30em', // 480px
    sm: '48em', // 768px
    md: '64em', // 1024px
    lg: '75em', // 1200px
    xl: '90em', // 1440px
  },
  colors: {
    'dark-blue': darkBlue,
    beige,
  },
  components: {
    Container: Container.extend({
      vars: (_, { size, fluid }) => ({
        root: {
          '--container-size': fluid
            ? '100%'
            : size !== undefined && size in CONTAINER_SIZES
              ? CONTAINER_SIZES[size]
              : rem(size),
        },
      }),
    }),
    Loader: Loader.extend({
      defaultProps: {
        loaders: { ...Loader.defaultLoaders, book: BookLoader },
      },
      vars: (_, props) => {
        if (props.size === 'xxl') {
          return {
            root: {
              '--loader-size': rem(90),
            },
          };
        }
        return { root: {} };
      },
    }),
  },
});
