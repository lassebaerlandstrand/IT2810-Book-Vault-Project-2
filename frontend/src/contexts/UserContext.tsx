import { createContext, ReactNode } from 'react';
import { Button, Flex, Loader, Stack, Text } from '@mantine/core';
import { User } from '@/generated/graphql';
import { makeUser } from '@/hooks/makeUser';
import { useUserHook } from '@/hooks/useUserHook';
import styles from './userContext.module.css';

interface UserContextProps {
  info: User;
  setUser: (user: User) => void;
}

const genNewUser = () => {
  localStorage.removeItem('userID');
  window.location.reload();
};

export const UserContext = createContext<UserContextProps | undefined>(undefined);

export const UserProvider = ({ children }: { children: ReactNode }) => {
  const userFunction = () => {
    const UUID = localStorage.getItem('userID');
    if (UUID) {
      const oldUser = useUserHook({ UUID });
      return oldUser;
    }
    const newUser = makeUser();
    if (newUser.user) {
      localStorage.setItem('userID', newUser.user.UUID);
    }
    return newUser;
  };

  const { user, loading, error } = userFunction();

  if (loading) {
    return (
      <Flex justify="center" align="center" className={styles.centeredOnPage}>
        <Stack align="center">
          <Text size="lg" component="h1">
            Fetching user data
          </Text>
          <Loader />
        </Stack>
      </Flex>
    );
  }

  if (error) {
    return (
      <Flex justify="center" align="center" className={styles.centeredOnPage}>
        <Stack align="center">
          <Text c="red" size="lg">
            Error fetching user data
          </Text>
        </Stack>
      </Flex>
    );
  }

  if (!user) {
    return (
      <Flex justify="center" align="center" className={styles.centeredOnPage}>
        <Stack align="center">
          <Text c="red" size="lg">
            Error fetching user data
          </Text>
          <Flex w="30rem">
            <Text size="lg" ta="center">
              This error may be caused by the user your userId reference having been removed from
              the database. Would you like to create a new user?
            </Text>
          </Flex>
          <Button onClick={genNewUser}>New user</Button>
        </Stack>
      </Flex>
    );
  }

  const setUser = () => {};

  return (
    <UserContext.Provider value={{ info: user as User, setUser }}>{children}</UserContext.Provider>
  );
};
