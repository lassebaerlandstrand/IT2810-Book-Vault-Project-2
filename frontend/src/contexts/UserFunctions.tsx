import { useContext } from 'react';
import { UserContext } from './UserContext';

/**
 * Custom hook to access the UserContext.
 * @returns {UserContextProps} The user context value containing user info and setter
 *     - info: The user info object
 *     - setUser: Function to set the user info
 * @throws {Error} If used outside of UserProvider
 */
export const useUser = () => {
  const context = useContext(UserContext);
  if (!context) {
    throw new Error('useUser must be used within a FavoritesProvider');
  }
  return context;
};
