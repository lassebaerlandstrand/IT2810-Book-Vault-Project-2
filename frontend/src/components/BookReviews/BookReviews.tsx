import { useState } from 'react';
import { useQuery } from '@apollo/client';
import { Button, Flex } from '@mantine/core';
import { useUser } from '@/contexts/UserFunctions';
import { Review as ReviewType } from '@/generated/graphql';
import { GET_BOOK_REVIEWS } from '@/graphql/queries/reviews';
import ReviewStack from '../ReviewStack/ReviewStack';

type BookReviewsProps = {
  bookId: string;
  top: React.RefObject<HTMLDivElement>;
};

/**
 * Displays and manages a paginated list of reviews for a specific book.
 */
const BookReviews = ({ bookId, top }: BookReviewsProps) => {
  const [page, setPage] = useState(0);
  const [isLastPage, setLastPage] = useState(true);
  const [displayReviews, setDisplayReviews] = useState<ReviewType[]>([]);

  const UUID: string = useUser().info.UUID;

  const { loading: loadingDisplayReviews, fetchMore } = useQuery(GET_BOOK_REVIEWS, {
    variables: { bookID: bookId, limit: 3, offset: page, userUUID: UUID },
    onCompleted: (data) => {
      setDisplayReviews((old) => [...old, ...(data.bookReviews.reviews as ReviewType[])]);
      setLastPage(data.bookReviews.pagination.isLastPage);
    },
  });

  const loadMoreReviews = () => {
    // Disable loading more when waiting for other reviews
    if (!loadingDisplayReviews) {
      fetchMore({
        variables: { offset: page + 1 },
      });
      setPage((currentPage) => currentPage + 1);
    }
  };

  const scrollToTop = () => {
    if (top.current) {
      window.scrollTo({
        top: top.current.offsetTop,
        behavior: 'smooth',
      });
    }
  };

  return (
    <>
      <ReviewStack reviews={displayReviews} type="profileReview" />

      <Flex justify="center" align="center" gap="lg">
        {!isLastPage ? (
          <Button variant="filled" onClick={loadMoreReviews}>
            Load more
          </Button>
        ) : (
          <></>
        )}
        {displayReviews.length > 5 ? (
          <Button variant="filled" onClick={scrollToTop}>
            Scroll to top
          </Button>
        ) : (
          <></>
        )}
      </Flex>
    </>
  );
};

export default BookReviews;
