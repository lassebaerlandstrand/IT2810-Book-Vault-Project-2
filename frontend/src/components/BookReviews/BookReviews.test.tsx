import { MockedProvider } from '@apollo/client/testing';
import { render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { MantineProvider } from '@mantine/core';
import { useUser } from '@/contexts/UserFunctions';
import { GET_BOOKS_REVIEWS } from '@/graphql/queries/reviews';
import { theme } from '@/theme';
import { dummyBook } from '../../../test-utils/testVars';
import BookReviews from './BookReviews';

vi.mock('@/contexts/UserFunctions', () => ({
  useUser: vi.fn(),
}));

const dummyUser = {
  info: {
    name: 'You',
    UUID: 'UUID1',
  },
};

const mocks = [
  {
    request: {
      query: GET_BOOKS_REVIEWS,
      variables: {
        bookID: dummyBook.id,
        limit: 3,
        offset: 0,
        userUUID: dummyUser.info.UUID,
      },
    },
    result: {
      data: {
        bookReviews: {
          reviews: [
            {
              UUID: '1',
              description: 'Great book!',
              rating: 5,
              at: '2024-01-01',
              user: { name: 'User1', UUID: 'user1' },
            },
            {
              UUID: '2',
              description: 'Not bad',
              rating: 3,
              at: '2024-01-02',
              user: { name: 'User2', UUID: 'user2' },
            },
            {
              UUID: '3',
              description: 'Could be better',
              rating: 2,
              at: '2024-01-03',
              user: { name: 'User3', UUID: 'user3' },
            },
          ],
          pagination: {
            isLastPage: false,
          },
        },
      },
    },
  },
];

const customRender = () => {
  const mockTopRef = { current: document.createElement('div') };

  return render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <MantineProvider theme={theme}>
        <MemoryRouter>
          <BookReviews bookId={dummyBook.id} top={mockTopRef} />
        </MemoryRouter>
      </MantineProvider>
    </MockedProvider>
  );
};

describe('Reviews Component', () => {
  beforeEach(() => {
    (useUser as jest.Mock).mockReturnValue(dummyUser);
  });

  test('When it has responses, show them', async () => {
    customRender();

    // Wait for the mock data to render in the component
    await waitFor(() => {
      expect(screen.getByText('Great book!')).toBeInTheDocument();
      expect(screen.getByText('Not bad')).toBeInTheDocument();
      expect(screen.getByText('Could be better')).toBeInTheDocument();
    });
  });

  it('matches snapshot', () => {
    const { asFragment } = customRender();

    const attributesToRemove = [
      ...document.body.querySelectorAll('div [id^="mantine"]'),
      ...document.body.querySelectorAll('div [for^="mantine"]'),
    ]; // Because Mantine uses random ids which causes snapshots to fail
    attributesToRemove.forEach((element) => {
      element.removeAttribute('for');
      element.removeAttribute('id');
      element.removeAttribute('class');
      element.removeAttribute('aria-describedby');
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
