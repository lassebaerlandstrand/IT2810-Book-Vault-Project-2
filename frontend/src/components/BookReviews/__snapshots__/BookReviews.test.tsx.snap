// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Reviews Component > matches snapshot 1`] = `
<DocumentFragment>
  <style
    data-mantine-styles="true"
  >
    :root{--mantine-font-family: Poppins, -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-font-family-headings: Poppins, -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-breakpoint-xs: 30em;--mantine-breakpoint-md: 64em;--mantine-breakpoint-xl: 90em;--mantine-breakpoint-xxs: 24em;--mantine-color-dark-blue-0: #155896;--mantine-color-dark-blue-1: #124b81;--mantine-color-dark-blue-2: #0f3f6c;--mantine-color-dark-blue-3: #0c3256;--mantine-color-dark-blue-4: #092641;--mantine-color-dark-blue-5: #06192b;--mantine-color-dark-blue-6: #051321;--mantine-color-dark-blue-7: #030d16;--mantine-color-dark-blue-8: #030a11;--mantine-color-dark-blue-9: #02070b;--mantine-color-beige-0: #fcfaf7;--mantine-color-beige-1: #f9f5ef;--mantine-color-beige-2: #f6f0e7;--mantine-color-beige-3: #f2eade;--mantine-color-beige-4: #eee4d3;--mantine-color-beige-5: #e9ddc8;--mantine-color-beige-6: #e2d1b5;--mantine-color-beige-7: #dac5a1;--mantine-color-beige-8: #d4bd94;--mantine-color-beige-9: #6a5f4a;}:root[data-mantine-color-scheme="dark"]{--mantine-color-dark-blue-text: var(--mantine-color-dark-blue-4);--mantine-color-dark-blue-filled: var(--mantine-color-dark-blue-8);--mantine-color-dark-blue-filled-hover: var(--mantine-color-dark-blue-9);--mantine-color-dark-blue-light: rgba(5, 19, 33, 0.15);--mantine-color-dark-blue-light-hover: rgba(5, 19, 33, 0.2);--mantine-color-dark-blue-light-color: var(--mantine-color-dark-blue-3);--mantine-color-dark-blue-outline: var(--mantine-color-dark-blue-4);--mantine-color-dark-blue-outline-hover: rgba(9, 38, 65, 0.05);--mantine-color-beige-text: var(--mantine-color-beige-4);--mantine-color-beige-filled: var(--mantine-color-beige-8);--mantine-color-beige-filled-hover: var(--mantine-color-beige-9);--mantine-color-beige-light: rgba(226, 209, 181, 0.15);--mantine-color-beige-light-hover: rgba(226, 209, 181, 0.2);--mantine-color-beige-light-color: var(--mantine-color-beige-3);--mantine-color-beige-outline: var(--mantine-color-beige-4);--mantine-color-beige-outline-hover: rgba(238, 228, 211, 0.05);}:root[data-mantine-color-scheme="light"]{--mantine-color-dark-blue-text: var(--mantine-color-dark-blue-filled);--mantine-color-dark-blue-filled: var(--mantine-color-dark-blue-6);--mantine-color-dark-blue-filled-hover: var(--mantine-color-dark-blue-7);--mantine-color-dark-blue-light: rgba(5, 19, 33, 0.1);--mantine-color-dark-blue-light-hover: rgba(5, 19, 33, 0.12);--mantine-color-dark-blue-light-color: var(--mantine-color-dark-blue-6);--mantine-color-dark-blue-outline: var(--mantine-color-dark-blue-6);--mantine-color-dark-blue-outline-hover: rgba(5, 19, 33, 0.05);--mantine-color-beige-text: var(--mantine-color-beige-filled);--mantine-color-beige-filled: var(--mantine-color-beige-6);--mantine-color-beige-filled-hover: var(--mantine-color-beige-7);--mantine-color-beige-light: rgba(226, 209, 181, 0.1);--mantine-color-beige-light-hover: rgba(226, 209, 181, 0.12);--mantine-color-beige-light-color: var(--mantine-color-beige-6);--mantine-color-beige-outline: var(--mantine-color-beige-6);--mantine-color-beige-outline-hover: rgba(226, 209, 181, 0.05);}
  </style>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 29.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 30em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 63.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 64em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 89.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 90em) {.mantine-hidden-from-xl {display: none !important;}}@media (max-width: 23.99375em) {.mantine-visible-from-xxs {display: none !important;}}@media (min-width: 24em) {.mantine-hidden-from-xxs {display: none !important;}}
  </style>
  <div
    class="_noResultWrapper_8d4f85 m_4081bf90 mantine-Group-root"
    style="--group-gap: var(--mantine-spacing-md); --group-align: center; --group-justify: center; --group-wrap: wrap;"
  >
    <svg
      class="tabler-icon tabler-icon-stars-off "
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M17.373 13.371l.076 -.154a.392 .392 0 0 1 .702 0l.907 1.831m.367 .39c.498 .071 1.245 .18 2.24 .324a.39 .39 0 0 1 .217 .665c-.326 .316 -.57 .553 -.732 .712m-.611 3.405a.39 .39 0 0 1 -.567 .411l-2.172 -1.138l-2.172 1.138a.392 .392 0 0 1 -.568 -.41l.415 -2.411l-1.757 -1.707a.389 .389 0 0 1 .217 -.665l1.601 -.232"
      />
      <path
        d="M6.2 19.817l-2.172 1.138a.392 .392 0 0 1 -.568 -.41l.415 -2.411l-1.757 -1.707a.389 .389 0 0 1 .217 -.665l2.428 -.352l1.086 -2.193a.392 .392 0 0 1 .702 0l1.086 2.193l2.428 .352a.39 .39 0 0 1 .217 .665l-1.757 1.707l.414 2.41a.39 .39 0 0 1 -.567 .411l-2.172 -1.138z"
      />
      <path
        d="M9.557 5.556l1 -.146l1.086 -2.193a.392 .392 0 0 1 .702 0l1.086 2.193l2.428 .352a.39 .39 0 0 1 .217 .665l-1.757 1.707l.414 2.41a.39 .39 0 0 1 -.014 .187m-4.153 -.166l-.744 .39a.392 .392 0 0 1 -.568 -.41l.188 -1.093"
      />
      <path
        d="M3 3l18 18"
      />
    </svg>
    <p
      class="mantine-focus-auto m_b6d8b162 mantine-Text-root"
      data-size="xl"
      style="--text-fz: var(--mantine-font-size-xl); --text-lh: var(--mantine-line-height-xl); margin-block: var(--mantine-spacing-xl); font-weight: 700;"
    >
      No reviews found
    </p>
  </div>
  <div
    class="m_8bffd616 mantine-Flex-root __m__-r50"
    style="gap: var(--mantine-spacing-lg); align-items: center; justify-content: center;"
  />
</DocumentFragment>
`;
